x/scheduler/keeper/keeper.go

```
@@ -163,14 +163,30 @@ func (k *Keeper) ScheduleTask(ctx context.Context, typename string, id types.Tas
 		return err
 	}
 
-	if key := getTaskScheduleKey(*task); key != nil {
+	return k.programTask(ctx, *task)
+}
+
+// programTask programs the task to its next run schedule, if any.
+func (k *Keeper) programTask(ctx context.Context, task types.Task) error {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	if key := getTaskScheduleKey(task); key != nil {
+		if err := sdkCtx.EventManager().EmitTypedEvent(&types.TaskScheduledEvent{
+			Id:       task.Id,
+			Typename: task.Typename,
+			At:       task.ScheduledFor,
+		}); err != nil {
+			return err
+		}
+
 		return k.tasksSchedule.Set(ctx, *key)
```

I think the event should be triggered after and it doesn't block setting the task if there's an error.

---

Yeah I'll move it you're right it's always a good practice to emit the event after, but in reality we are operating on a cached context so if either fails the whole context is dropped

---

yeah but probably we need to just log this error to avoid this scenario, no?

---

If we allow the event emission to fail by silencing the error we accept inconsistent block events, I prefer to ensure that the events are aligned with state changes

---

yeah, makes sense! I was wondering if it would be good or not to block state transactions / txs from happening because of failed events, since the cached context is dropped if the error happen.